File: sdk/python/testvectors/BlockFactory.py
from symbolchain import sc
from symbolchain.ByteArray import ByteArray
from symbolchain.CryptoTypes import Hash256, PublicKey
from symbolchain.RuleBasedTransactionFactory import RuleBasedTransactionFactory
from symbolchain.symbol.Network import Address


class ProofGamma(ByteArray):
	"""Represents a 256-bit vrf proof gamma."""

	SIZE = 32

	def __init__(self, proof_gamma):
		"""Creates a vrf proof gamma from bytes or a hex string."""
		super().__init__(self.SIZE, proof_gamma, ProofGamma)

	def __repr__(self):
		return f'ProofGamma(\'{str(self)}\')'


class ProofVerificationHash(ByteArray):
	"""Represents a 128-bit vrf proof verification hash."""

	SIZE = 16

	def __init__(self, proof_verification_hash):
		"""Creates a proof verification hash from bytes or a hex string."""
		super().__init__(self.SIZE, proof_verification_hash, ProofVerificationHash)

	def __repr__(self):
		return f'ProofVerificationHash(\'{str(self)}\')'


class ProofScalar(ByteArray):
	"""Represents a 256-bit vrf proof scalar."""

	SIZE = 32

	def __init__(self, proof_scalar):
		"""Creates a vrf proof scalar from bytes or a hex string."""
		super().__init__(self.SIZE, proof_scalar, ProofScalar)

	def __repr__(self):
		return f'ProofScalar(\'{str(self)}\')'


class BlockFactory:
	"""Factory for creating Symbol blocks."""

	def __init__(self, network, type_rule_overrides=None):
		"""Creates a factory for the specified network."""
		self.factory = self._build_rules(type_rule_overrides)
		self.network = network

	def _create_and_extend(self, block_descriptor, factory_class):
		block = self.factory.create_from_factory(factory_class.create_by_name, {
			**block_descriptor,
			'network': self.network.identifier
		})

		return block

	def create(self, block_descriptor):
		"""Creates a block from a block descriptor."""
		return self._create_and_extend(block_descriptor, sc.BlockFactory)

	@staticmethod
	def _symbol_type_converter(value):
		if isinstance(value, Address):
			return sc.UnresolvedAddress(value.bytes)

		return None

	@staticmethod
	def _build_rules(type_rule_overrides):
		factory = RuleBasedTransactionFactory(sc, BlockFactory._symbol_type_converter, type_rule_overrides)
		factory.autodetect()

		factory.add_struct_parser('VrfProof')

		sdk_type_block_mapping = {
			'UnresolvedAddress': Address,
			'Address': Address,
			'Hash256': Hash256,
			'PublicKey': PublicKey,
			'ProofGamma': ProofGamma,
			'ProofVerificationHash': ProofVerificationHash,
			'ProofScalar': ProofScalar
		}
		for name, typename in sdk_type_block_mapping.items():
			factory.add_pod_parser(name, typename)

		for name in ['BlockType', 'UnresolvedAddress']:
			factory.add_array_parser(name)

		return factory

File: sdk/javascript/vectors/BlockFactory.js
import ByteArray from '../src/ByteArray.js';
import { Hash256, PublicKey } from '../src/CryptoTypes.js';
import RuleBasedTransactionFactory from '../src/RuleBasedTransactionFactory.js';
import {
	Address,
	/* eslint-disable no-unused-vars */
	Network
	/* eslint-enable no-unused-vars */
} from '../src/symbol/Network.js';
import * as sc from '../src/symbol/models.js';

/**
 * Represents a 256-bit vrf proof gamma.
 */
class ProofGamma extends ByteArray {
	static SIZE = 32;

	/**
	 * Creates a vrf proof gamma from bytes or a hex string.
	 * @param {Uint8Array|string} proofGamma Input string or byte array.
	 */
	constructor(proofGamma) {
		super(ProofGamma.SIZE, proofGamma);
	}
}

/**
 * Represents a 128-bit vrf proof verification hash.
 */
class ProofVerificationHash extends ByteArray {
	static SIZE = 16;

	/**
	 * Creates a proof verification hash from bytes or a hex string.
	 * @param {Uint8Array|string} proofVerificationHash Input string or byte array.
	 */
	constructor(proofVerificationHash) {
		super(ProofVerificationHash.SIZE, proofVerificationHash);
	}
}

/**
 * Represents a 256-bit vrf proof scalar.
 */
class ProofScalar extends ByteArray {
	static SIZE = 32;

	/**
	 * Creates a vrf proof scalar from bytes or a hex string.
	 * @param {Uint8Array|string} proofScalar Input string or byte array.
	 */
	constructor(proofScalar) {
		super(ProofScalar.SIZE, proofScalar);
	}
}

/**
 * Factory for creating Symbol blocks.
 */
export default class BlockFactory {
	/**
	 * Creates a factory for the specified network.
	 * @param {Network} network Symbol network.
	 * @param {Map<string, function>|undefined} typeRuleOverrides Type rule overrides.
	 */
	constructor(network, typeRuleOverrides = undefined) {
		this.factory = BlockFactory._buildRules(typeRuleOverrides); // eslint-disable-line no-underscore-dangle
		this.network = network;
	}

	_createAndExtend(blockDescriptor, FactoryClass) {
		const block = this.factory.createFromFactory(FactoryClass.createByName, {
			...blockDescriptor,
			network: this.network.identifier
		});

		return block;
	}

	/**
	 * Creates a block from a block descriptor.
	 * @param {object} blockDescriptor Block descriptor.
	 * @returns {object} Newly created block.
	 */
	create(blockDescriptor) {
		return this._createAndExtend(blockDescriptor, sc.BlockFactory);
	}

	static _symbolTypeConverter(value) {
		if (value instanceof Address)
			return new sc.UnresolvedAddress(value.bytes);

		return undefined;
	}

	static _buildRules(typeRuleOverrides) {
		const factory = new RuleBasedTransactionFactory(sc, this._symbolTypeConverter, typeRuleOverrides);
		factory.autodetect();

		['BlockType', 'NetworkType'].forEach(name => { factory.addEnumParser(name); });

		factory.addStructParser('VrfProof');

		const sdkTypeMapping = {
			UnresolvedAddress: Address,
			Address,
			Hash256,
			PublicKey,
			ProofGamma,
			ProofVerificationHash,
			ProofScalar
		};
		Object.keys(sdkTypeMapping).forEach(name => { factory.addPodParser(name, sdkTypeMapping[name]); });

		['BlockType', 'UnresolvedAddress'].forEach(name => {
			factory.addArrayParser(name);
		});

		return factory;
	}
}
